### PR TITLE
LP-2821 - Integrate Webinar List view for Admin Dashboard

### PR DESCRIPTION
**Pull request by aightmunam**
[LP-2821](https://philanthropyu.atlassian.net/browse/LP-2821)

[Theme PR](https://github.com/OmnipreneurshipAcademy/adg-edx-theme/pull/500)

***

#### Description:

- Add Webinar list view for Admin Dashboard
- Write API endpoints for cancellation and cloning of webinars
- Add saved, upcoming and cancelled webinars to the webinar list page
- Add unit tests

### CODE REVIEW

**Comment by omnipreneurshipA** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/2836/workflows/workflows_id/jobs/2836/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/admin_dashboard/webinars/constants.py (100%)
openedx/adg/admin_dashboard/webinars/helpers.py (100%)
openedx/adg/admin_dashboard/webinars/tests/test_helpers.py (100%)
openedx/adg/admin_dashboard/webinars/tests/test_views.py (100%)
openedx/adg/admin_dashboard/webinars/urls.py (100%)
openedx/adg/admin_dashboard/webinars/views.py (100%)
openedx/adg/lms/webinars/admin.py (0.0%): Missing lines 54-55
openedx/adg/lms/webinars/api/constants.py (100%)
openedx/adg/lms/webinars/api/tests/test_api_views.py (100%)
openedx/adg/lms/webinars/api/urls.py (100%)
openedx/adg/lms/webinars/api/views.py (100%)
openedx/adg/lms/webinars/models.py (96.0%): Missing lines 196
openedx/adg/lms/webinars/tests/test_models.py (100%)
-------------
Total:   199 lines
Missing: 3 lines
Coverage: 98%
-------------
```

File: openedx/adg/lms/webinars/models.py [hunk 1]
```diff
@@ -331,13 +365,41 @@ def get_webinar_update_recipients_emails(self):
...
+    @classmethod
+    def clone(cls, webinar_to_clone):
+        """
+        Clone the given webinar as draft
+
+        Arguments:
+            webinar_to_clone (Webinar): Webinar that we want to clone
+        """
+        old_co_hosts = webinar_to_clone.co_hosts.all()
+        old_panelists = webinar_to_clone.panelists.all()
```
Comment (asadali145) [inline openedx/adg/lms/webinars/models.py:377]: No need to prepend `old`.

File: openedx/adg/lms/webinars/models.py [hunk 2]
```diff
@@ -331,13 +365,41 @@ def get_webinar_update_recipients_emails(self):
...
+
+        webinar_to_clone.pk = None
+        webinar_to_clone.id = None
```
Comment (asadali145) [inline openedx/adg/lms/webinars/models.py:380]: `webinar_to_clone .pk = None` would do it. No need to make id=None

File: openedx/adg/lms/webinars/models.py [hunk 3]
```diff
@@ -331,13 +365,41 @@ def get_webinar_update_recipients_emails(self):
...
+
+        webinar_to_clone.is_published = False
+        webinar_to_clone.is_cancelled = False
+        webinar_to_clone.modified_by = None
+        webinar_to_clone.modified = now()
```
Comment (asadali145) [inline openedx/adg/lms/webinars/models.py:385]: Why we are setting date for modified but not the modified_by? We should actually use the request.user and set modified by and created by.

**Review by asadali145 — CHANGES_REQUESTED** [review]
Great work! It is good to see code improvements. Here are a  few suggestions.

**Comment by asadali145** [comment]
@aightmunam Are we going to handle delete webinars from list view in some other ticket?
CC: @javeriaejaz-arbisoft

File: openedx/adg/admin_dashboard/webinars/views.py [hunk 4]
```diff
@@ -1,3 +1,88 @@
...
+    def _get_selected_number_of_webinars(self, webinar_queryset, limit_by=SHOW_LIMITED_WEBINARS):
+        """
+        Get all webinars if `SHOW_ALL_ITEMS` is selected for a particular type of webinars, else limited
+        number of webinars are selected
+
+        Args:
+            webinar_queryset (QuerySet): QuerySet containing all the webinars
+            limit_by (str): Flag to check if all webinars should be selected or not
+        """
+        return webinar_queryset[:LIMIT] if limit_by != SHOW_ALL_WEBINARS else webinar_queryset
```
[the remark this answers is not in the document]
Comment (aightmunam) [reply openedx/adg/admin_dashboard/webinars/views.py:41]: Did it this way for a reason. Incase a user manually changes the url, the default will always be LIMIT. Only if SHOW_ALL_WEBINARS is given as query param, will it show all webinars.

File: openedx/adg/lms/webinars/api/views.py [hunk 5]
```diff
@@ -71,3 +74,59 @@ def post(self, request, pk, action):
...
+            HttpResponse: On success returns response with status code 200 otherwise
+            throw HTTP error code
+        """
+        try:
+            webinar_to_clone = Webinar.objects.get(pk=pk)
+        except ObjectDoesNotExist:
+            return HttpResponseNotFound(WEBINAR_NOT_FOUND_ERROR_MESSAGE)
+
+        Webinar.clone(webinar_to_clone)
+        return Response({}, status=HTTP_200_OK)
```
[the remark this answers is not in the document]
Comment (aightmunam) [reply openedx/adg/lms/webinars/api/views.py:104]: Do you mean returning Response vs HTTPResponse?

[the remark this answers is not in the document]
Comment (aightmunam) [reply openedx/adg/lms/webinars/api/views.py:104]: Just following the pattern in other API views already written.

File: openedx/adg/lms/webinars/models.py (hunk 2 above)
Comment (aightmunam) [reply openedx/adg/lms/webinars/models.py:380]: We do need to make id none as we are inheriting from TimeStampedModel here. Take a look here: https://docs.djangoproject.com/en/dev/topics/db/queries/#copying-model-instances

File: openedx/adg/lms/webinars/models.py (hunk 3 above)
Comment (aightmunam) [reply openedx/adg/lms/webinars/models.py:385]: It is set automatically by save method. It has already been overridden before to cater for this.

File: openedx/adg/lms/webinars/api/views.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (asadali145) [reply openedx/adg/lms/webinars/api/views.py:104]: Yes Response vs HttpResponse. It shouldn't be like this. Can we do something for this now?

[the remark this answers is not in the document]
Comment (aightmunam) [reply openedx/adg/lms/webinars/api/views.py]: Used `get_object_or_404` from rest_framework.generics

File: openedx/adg/admin_dashboard/webinars/constants.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
+"""
+Constants for webinars app
+"""
+from django.utils.translation import gettext as _
+
+SHOW_ALL_WEBINARS = _('all')
+SHOW_LIMITED_WEBINARS = _('limit')
+
+LIMIT = 3
```
Comment (asadali145) [inline openedx/adg/admin_dashboard/webinars/constants.py:9]: Can we have a more meaningful name for this one?

**Comment by omnipreneurshipA** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/2837/workflows/workflows_id/jobs/2837/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/admin_dashboard/webinars/constants.py (100%)
openedx/adg/admin_dashboard/webinars/helpers.py (100%)
openedx/adg/admin_dashboard/webinars/tests/test_helpers.py (100%)
openedx/adg/admin_dashboard/webinars/tests/test_views.py (100%)
openedx/adg/admin_dashboard/webinars/urls.py (100%)
openedx/adg/admin_dashboard/webinars/views.py (100%)
openedx/adg/lms/webinars/admin.py (0.0%): Missing lines 54-55
openedx/adg/lms/webinars/api/constants.py (100%)
openedx/adg/lms/webinars/api/tests/test_api_views.py (100%)
openedx/adg/lms/webinars/api/urls.py (100%)
openedx/adg/lms/webinars/api/views.py (100%)
openedx/adg/lms/webinars/models.py (96.0%): Missing lines 196
openedx/adg/lms/webinars/tests/test_models.py (100%)
-------------
Total:   194 lines
Missing: 3 lines
Coverage: 98%
-------------
```

**Review by asadali145 — APPROVED** [review]
Great Work! A small suggestion rest looks good.

**Comment by omnipreneurshipA** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/2842/workflows/workflows_id/jobs/2842/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/admin_dashboard/webinars/api/tests/test_api_views.py (100%)
openedx/adg/admin_dashboard/webinars/api/urls.py (100%)
openedx/adg/admin_dashboard/webinars/api/views.py (100%)
openedx/adg/admin_dashboard/webinars/constants.py (100%)
openedx/adg/admin_dashboard/webinars/helpers.py (100%)
openedx/adg/admin_dashboard/webinars/tests/test_helpers.py (100%)
openedx/adg/admin_dashboard/webinars/tests/test_views.py (100%)
openedx/adg/admin_dashboard/webinars/urls.py (100%)
openedx/adg/admin_dashboard/webinars/views.py (100%)
openedx/adg/conftest.py (100%)
openedx/adg/lms/webinars/admin.py (0.0%): Missing lines 54-55
openedx/adg/lms/webinars/api/views.py (100%)
openedx/adg/lms/webinars/models.py (96.0%): Missing lines 196
openedx/adg/lms/webinars/tests/test_models.py (100%)
-------------
Total:   230 lines
Missing: 3 lines
Coverage: 98%
-------------
```

**Review by AliAdnanSohail — APPROVED** [review]

**Review by asadali145 — APPROVED** [review]

**Review by faizan-philu — APPROVED** [review]

**Pull request merged**